### PR TITLE
CDAP-7540 CDAP-7541 DataTracer API to be used by applications to log the preview data and PreviewStore Implementation

### PR DESCRIPTION
**Pull request by yaojiefeng**
JIRA: https://issues.cask.co/browse/CDAP-7540
          https://issues.cask.co/browse/CDAP-7541
Build: http://builds.cask.co/browse/CDAP-DUT5012-1
Docs: https://wiki.cask.co/pages/viewpage.action?pageId=4360859

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/NoopDebugLogger.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+public class NoopDebugLogger implements DebugLogger {
+
+  @Override
+  public void info(String propertyName, Object propertyValue) {
+    // no-op
+  }
+
+  @Override
+  public String getName() {
+    return null;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/NoopDebugLogger.java]: shouldn't be null. Should be the name of the logger.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java [hunk 2]
```diff
@@ -93,9 +98,11 @@ protected AbstractContext(Program program, ProgramOptions programOptions,
                             Set<String> datasets, DatasetFramework dsFramework, TransactionSystemClient txClient,
                             DiscoveryServiceClient discoveryServiceClient, boolean multiThreaded,
                             @Nullable MetricsCollectionService metricsService, Map<String, String> metricsTags,
-                            SecureStore secureStore, SecureStoreManager secureStoreManager) {
+                            SecureStore secureStore, SecureStoreManager secureStoreManager,
+                            @Nullable PreviewId previewId, DebugLoggerFactory debugLoggerFactory) {
     this(program, programOptions, datasets, dsFramework, txClient,
-         discoveryServiceClient, multiThreaded, metricsService, metricsTags, secureStore, secureStoreManager, null);
+         discoveryServiceClient, multiThreaded, metricsService, metricsTags, secureStore, secureStoreManager, null,
+         previewId, debugLoggerFactory);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: Better use optional injection than constructor injection

https://github.com/google/guice/wiki/Injections

Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: Also, how's the previewId is being used? Why it is not the same the application ID? Seems like this change is quite intrusive.

Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: Shall we use ApplicationId instead of PreviewId? The PreviewId will be generated from PreviewManager and passed using system arguments, which is not implemented yet.

Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: Is it? I thought under preview mode, the preview id is generated and is used as the application id? @sagarkapare can you clarify?

Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: @chtyim yes. In preview mode, we generate the preview id and use it as application id. Relevant code is - 

``` java
    final PreviewId previewId = new PreviewId(namespaceId.getNamespace(), RunIds.generate().getId());
    AppRequest<?> appRequest = GSON.fromJson(config, AppRequest.class);
    ArtifactSummary artifactSummary = appRequest.getArtifact();
    NamespaceId artifactNamespace =
      ArtifactScope.SYSTEM.equals(artifactSummary.getScope()) ? NamespaceId.SYSTEM : namespaceId;
    ArtifactId artifactId =
      new ArtifactId(artifactNamespace.getNamespace(), artifactSummary.getName(), artifactSummary.getVersion());

    // if we don't null check, it gets serialized to "null"
    String configString = appRequest.getConfig() == null ? null : GSON.toJson(appRequest.getConfig());
    try {
      applicationLifecycleService.deployApp(namespaceId.toId(), previewId.getPreview(), artifactId.toId(), configString,
                                            new ProgramTerminator() {
                                              @Override
                                              public void stop(Id.Program programId) throws Exception {

                                              }
                                            });
    } catch (Exception e) {
      previewStatusMap.put(previewId, new PreviewStatus(PreviewStatus.Status.DEPLOY_FAILED, e));
      throw Throwables.propagate(e);
    }
```

Seems like `PreviewId` is redundant and we can use `ApplicationId` instead.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java [hunk 3]
```diff
@@ -84,6 +89,7 @@
   private final long logicalStartTime;
   private final SecureStore secureStore;
   private final Transactional transactional;
+  @Inject(optional = true) private final DebugLoggerFactory debugLoggerFactory = new NoopDebugLoggerFactory();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: don't use final field with field injection. Better use non-final field with method injection.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java [hunk 4]
```diff
@@ -355,4 +361,9 @@ public void execute(TxRunnable runnable) throws TransactionFailureException {
   public void execute(int timeoutInSeconds, TxRunnable runnable) throws TransactionFailureException {
     transactional.execute(timeoutInSeconds, runnable);
   }
+
+  public DebugLogger getLogger(String loggerName) {
+    return debugLoggerFactory.getLogger(loggerName,
+                                        new ApplicationId(program.getNamespaceId(), program.getApplicationId()));
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: Why construct a new one? You can just call `program.getId().getParent()`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+  private PreviewDataset getPreviewDataset (DatasetContext datasetContext)
+    throws IOException, DatasetManagementException {
+    Table table =  DatasetsUtil.getOrCreateDataset(datasetContext, dsFramework, PREVIEW_TABLE_ID, Table.class.getName(),
+                                                   DatasetProperties.EMPTY);
+    return new PreviewDataset(table);
+  }
+
+  private void truncate(DatasetAdmin admin) throws IOException {
+    if (admin != null) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java:139]: is admin nullable?

Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java:139]: The admin is get from `dsFrameWork.getAdmin(name)`, which will be null if dataset instances of the provided name does not exist.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGenerator.java [hunk 6]
```diff
@@ -261,13 +261,13 @@ private void generateConstructor(TypeToken<? extends HttpServiceHandler> delegat
...
   private void generateLogger(Type classType, ClassWriter classWriter) {
-    // private static final Logger LOG = LoggerFactory.getLogger(classType);
+    // private static final Logger LOG = LoggerFactory.getTracer(classType);
     classWriter.visitField(Opcodes.ACC_PRIVATE | Opcodes.ACC_STATIC | Opcodes.ACC_FINAL, "LOG",
                            Type.getType(Logger.class).getDescriptor(), null, null);
     Method init = Methods.getMethod(void.class, "<clinit>");
     GeneratorAdapter mg = new GeneratorAdapter(Opcodes.ACC_STATIC, init, null, null, classWriter);
     mg.visitLdcInsn(classType);
-    mg.invokeStatic(Type.getType(LoggerFactory.class), Methods.getMethod(Logger.class, "getLogger", Class.class));
+    mg.invokeStatic(Type.getType(LoggerFactory.class), Methods.getMethod(Logger.class, "getTracer", Class.class));
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGenerator.java]: This is unrelated change.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewDataset.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  private static final byte[] DATA_RECORD_TYPE = Bytes.toBytes("d");
+
+  private final Table table;
+
+  PreviewDataset(Table table) {
+    super(PREVIEW_TABLE_NAME, table);
+    this.table = table;
+  }
+
+  void put(ApplicationId applicationId, String tracerName, String propertyName, Object value) {
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewDataset.java]: Please add javadoc explaining how the row keys are formed and schema.

File: cdap-watchdog/src/main/resources/logging/schema/LoggingEvent.avsc [hunk 8]
```diff
@@ -44,7 +44,7 @@
         "type" : [ "null", "string" ],
         "default" : null
     }, {
-        "name" : "loggerName",
+        "name" : "tracerName",
```
Comment (sagarkapare) [inline cdap-watchdog/src/main/resources/logging/schema/LoggingEvent.avsc]: unintended change?

File: cdap-api/src/main/java/co/cask/cdap/api/RuntimeContext.java [hunk 9]
```diff
@@ -50,4 +51,11 @@
    * @return an {@link Admin} to perform admin operations.
    */
   Admin getAdmin();
+
+  /**
+   * @param tracerName the name of the logger using which the debug information will be logged
+   *
+   * @return an {@link DataTracer} to perform log operations.
+   */
+  DataTracer getTracer(String tracerName);
```
Comment (sagarkapare) [inline cdap-api/src/main/java/co/cask/cdap/api/RuntimeContext.java]: I think `getDataTracer` would be more correct name since its called on `RuntimeContext`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGenerator.java (hunk 6 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGenerator.java]: Oh, my bad, it was changed by intellij automatically when renaming

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/AppFabricServiceRuntimeModule.java [hunk 10]
```diff
@@ -318,6 +320,7 @@ protected void configure() {
       bind(RuntimeStore.class).to(DefaultStore.class);
       bind(ArtifactStore.class).in(Scopes.SINGLETON);
       bind(ProgramLifecycleService.class).in(Scopes.SINGLETON);
+      bind(PreviewStore.class).to(DefaultPreviewStore.class);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/AppFabricServiceRuntimeModule.java]: Why add to this module? Shouldn't it in the preview module only? Adding it here means it will affect non-preview JVM as well.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java [hunk 11]
```diff
@@ -355,4 +360,14 @@ public void execute(TxRunnable runnable) throws TransactionFailureException {
...
     transactional.execute(timeoutInSeconds, runnable);
   }
+
+  @Override
+  public DataTracer getDataTracer(String dataTracerName) {
+    return dataTracerFactory.getTracer(dataTracerName, program.getId().getParent());
+  }
+
+  @Inject(optional = true)
+  private void setDataTracerFactory(@Named("DataTracerFactory") DataTracerFactory dataTracerFactory) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: Why need to use a named injection?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java (hunk 5 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java:139]: Then the parameter should be annotated with `@Nullable`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+        }
+      });
+    } catch (TransactionFailureException e) {
+      throw Transactions.propagate(e);
+    }
+  }
+
+  @VisibleForTesting
+  @Override
+  public void clear() throws IOException, DatasetManagementException {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java]: For method that is visible for testing, it shouldn't be `public`, but instead should be package protected. Also, I don't see any unit-test in this PR, how do you know this method is needed to be visible for testing?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java (hunk 11 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: Do not quite understand the concept of optional and method injection. Just followed the examples. If we use method injection, do we need to call this method or the variable will be injected without calling?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java (hunk 12 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java]: There is a unit-test named `DefaultPreviewStoreTest.java` in this PR. It is needed to truncate the table when the unit test is done

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java (hunk 11 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: You don't need to call this method and the Guice framework will call it. However, you shouldn't have the `@Named`, unless you have different bindings for the `DataTracerFactory` type.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java (hunk 12 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java]: Why? Usually unit-test doesn't require cleanup after run.

Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java]: Sorry, my bad for not getting things clear. It is needed to get called in before each unit test is run and make sure the `DefaultPreviewStore` is empty before each unit test.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java (hunk 11 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: I see. Thanks for the explanation!

**Comment by yaojiefeng** [comment]
@chtyim @sagarkapare Addressed comments. Please take a look, thank you!

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewDataset.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+   */
+  void put(ApplicationId applicationId, String tracerName, String propertyName, Object value) {
+    MDSKey mdsKey = new MDSKey.Builder().add(applicationId.getNamespace())
+      .add(applicationId.getApplication()).add(tracerName).add(COUNT_RECORD_TYPE).build();
+    long recordCount = table.incrementAndGet(mdsKey.getKey(), COUNT_RECORD_TYPE, 1L);
+
+    mdsKey = new MDSKey.Builder().add(applicationId.getNamespace())
+      .add(applicationId.getApplication()).add(tracerName).add(DATA_RECORD_TYPE).add(recordCount).build();
+
+    byte[][] columns = new byte[][] { LOGGER, PROPERTY, VALUE };
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewDataset.java]: The `columns` never change. Define it as a constant.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewDataset.java [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+   *
+   * @param applicationId application id of the preview run.
+   * @param tracerName the name of the {@link co.cask.cdap.api.preview.DataTracer}.
+   * @param propertyName the property of the data.
+   * @param value the value of the data.
+   */
+  void put(ApplicationId applicationId, String tracerName, String propertyName, Object value) {
+    MDSKey mdsKey = new MDSKey.Builder().add(applicationId.getNamespace())
+      .add(applicationId.getApplication()).add(tracerName).add(COUNT_RECORD_TYPE).build();
+    long recordCount = table.incrementAndGet(mdsKey.getKey(), COUNT_RECORD_TYPE, 1L);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewDataset.java]: This would be quite inefficient way to form the row key, as each put becomes two writes + one read.

Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewDataset.java]: Yes. Since we want to have multiple writes to the PreviewDataset with the applicationId and TracerName. I also tried to add the current timestamp to the mdskey but that does not work since some put will happen at the same time. Can I generate an UUID and add it to the row key? This will destroy the order of results from`get` when we scan the table.

Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewDataset.java]: Do we have one instance of `PreviewDataset` per preview run? If that's guaranteed, then we can simply have an `AtomicInteger` to increment

Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewDataset.java]: I think we will get the `PreviewDataset` for every `put` operation using DatasetContext like `getPreviewDataset(context).put`, so we will have a new `AtomicInteger` which is not what we want. Making it `static`will not make sense since we are keeping incrementing it for each preview run

Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewDataset.java]: We can have the `DefaultPreviewStore` holds the `AtomicInteger` and inject it as Singleton. In this case, we need to synchronize all operations in `DefaultPreviewStore`

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/DataTracerFactory.java [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+ * DataTracerFactory will be injected in the Program context classes. This may not be directly used by Applications.
+ */
+public interface DataTracerFactory {
+
+  /**
+   * Get the {@link DataTracer} used to put the debug data.
+   * @param tracerName the name of the tracer with which the data to be associated
+   * @return the instance of the DataTracer
+   */
+  DataTracer getTracer(String tracerName, ApplicationId applicationId);
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/DataTracerFactory.java]: `getDataTracer(ApplicationId applicationId, String tracerName)`?

File: cdap-api/src/main/java/co/cask/cdap/api/preview/DataTracer.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.api.preview;
+
+/**
+ * Interface used by the CDAP applications to log the debug data.
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/preview/DataTracer.java:19]: We need to expand the javadoc of this class and the `getDataTracer` method in the `RuntimeContext` to clearly state what's the usage of this `DataTracer` and what's the effect. Don't need to do it in this PR, but open a blocking JIRA for 4.0 to make sure it get fixed before the release.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/DataTracerFactory.java [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+ * DataTracerFactory will be injected in the Program context classes. This may not be directly used by Applications.
+ */
+public interface DataTracerFactory {
+
+  /**
+   * Get the {@link DataTracer} used to put the debug data.
+   * @param tracerName the name of the tracer with which the data to be associated
+   * @return the instance of the DataTracer
+   */
+  DataTracer getDataTracer(String tracerName, ApplicationId applicationId);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/DataTracerFactory.java]: I would put `ApplicationId` before `tracerName`, as that's the hierarchy is.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultDataTracer.java [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+ * the License.
+ */
+package co.cask.cdap.internal.app.preview;
+
+import co.cask.cdap.api.preview.DataTracer;
+import co.cask.cdap.app.store.preview.PreviewStore;
+import co.cask.cdap.proto.id.ApplicationId;
+
+/**
+ * Default implementation of {@link DataTracer}
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultDataTracer.java]: Expand the javadoc a bit, saying that data are being preserved using `PreviewStore`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultDataTracer.java [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+ */
+public class DefaultDataTracer implements DataTracer {
+
+  private final String tracerName;
+  private final ApplicationId applicationId;
+  private final PreviewStore previewStore;
+
+  public DefaultDataTracer(String tracerName, ApplicationId applicationId, PreviewStore previewStore) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultDataTracer.java]: revert the order of `tracerName` and `applicationId`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultDataTracer.java [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+package co.cask.cdap.internal.app.preview;
+
+import co.cask.cdap.api.preview.DataTracer;
+import co.cask.cdap.app.store.preview.PreviewStore;
+import co.cask.cdap.proto.id.ApplicationId;
+
+/**
+ * Default implementation of {@link DataTracer}
+ */
+public class DefaultDataTracer implements DataTracer {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultDataTracer.java]: This should be a package private class, since only `DefaultDataTracerFactory` should use  it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/NoopDataTracer.java [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+ * the License.
+ */
+package co.cask.cdap.internal.app.preview;
+
+import co.cask.cdap.api.preview.DataTracer;
+
+/**
+ * Implementation of the {@link DataTracer} which logs nothing.
+ */
+public class NoopDataTracer implements DataTracer {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/NoopDataTracer.java]: package private.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+ * Default implementation of the {@link PreviewStore} that stores data in in-memory table
+ */
+public class DefaultPreviewStore implements PreviewStore {
+
+  private static final DatasetId PREVIEW_TABLE_ID = NamespaceId.SYSTEM.dataset("preview.table");
+  private static final Gson GSON = new Gson();
+  private static final byte[] TRACER = Bytes.toBytes("l");
+  private static final byte[] PROPERTY = Bytes.toBytes("p");
+  private static final byte[] VALUE = Bytes.toBytes("v");
+  private static final byte[][] columns = new byte[][] {TRACER, PROPERTY, VALUE };
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java]: constant name should be all caps `COLUMNS`

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStoreTest.java [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+    store.put(firstApplicationId, "myanothertracer", "key2", 3);
+
+    // put data for the second application
+    store.put(secondApplicationId, "mytracer", "key1", "value1");
+
+    // get the data for first application and logger name "mytracer"
+    Map<String, List<String>> firstApplicationData = store.get(firstApplicationId, "mytracer");
+    // key1 and key2 are two keys inserted for the first application.
+    Assert.assertEquals(2, firstApplicationData.size());
+    Assert.assertEquals(Arrays.asList("\"value1\"", "2"), firstApplicationData.get("key1"));
```
Comment (chtyim) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStoreTest.java]: Why `value1` has `"` around it while the `put` call doesn't (line 64)?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/BasicSparkClientContext.java [hunk 24]
```diff
@@ -159,6 +160,11 @@ public Admin getAdmin() {
   }
 
   @Override
+  public DataTracer getDataTracer(String dataTracerName) {
```
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/BasicSparkClientContext.java:163]: Is the `DataTracer` returned serializable? If not, then it can't be used inside spark function through closure.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStoreTest.java (hunk 23 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStoreTest.java]: It has `"` around it because in `put` operation, we are using Gson.toJson() to store the value as we do not know the actual type of the value. And in `get` operation we are returning the jsonstr.

File: cdap-api/src/main/java/co/cask/cdap/api/preview/DataTracer.java (hunk 16 above)
Comment (yaojiefeng) [reply cdap-api/src/main/java/co/cask/cdap/api/preview/DataTracer.java:19]: Created JIRA: https://issues.cask.co/browse/CDAP-7610

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java [hunk 25]
```diff
@@ -75,4 +75,5 @@
   public static final String PRINCIPAL = "principal";
 
   public static final String KEYTAB_URI = "keytabURI";
+
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java]: Seems like an unnecessary change.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/preview/SparkDataTracer.java [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+   * Constructor. It delegates service discovery to the current {@link SparkRuntimeContext}.
+   */
+  public SparkDataTracer(SparkRuntimeContext sparkRuntimeContext, String tracerName) {
+    this.sparkRuntimeContext = sparkRuntimeContext;
+    this.tracerName = tracerName;
+  }
+
+  @Override
+  public void info(String propertyName, Object propertyValue) {
+    sparkRuntimeContext.getDataTracer(tracerName).info(propertyName, propertyValue);
```
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/preview/SparkDataTracer.java]: So each call will create a new `DataTracer`? Maybe better to create it once in the constructor?

**Comment by chtyim** [comment]
Mostly looks good, just few comments left.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java [hunk 27]
```diff
@@ -0,0 +1,200 @@
...
+import javax.annotation.Nullable;
+
+/**
+ * Default implementation of the {@link PreviewStore} that stores data in in-memory table
+ */
+public class DefaultPreviewStore implements PreviewStore {
+
+  private static final DatasetId PREVIEW_TABLE_ID = NamespaceId.SYSTEM.dataset("preview.table");
+  private static final Gson GSON = new Gson();
+  private static final byte[] TRACER = Bytes.toBytes("l");
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java]: should this be `t` (left over from logger?) ?

Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java]: Yes, missed that, thanks!

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/preview/SparkDataTracer.java [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+  * It has no-op for serialize/deserialize operation, with all operations delegated to the {@link DataTracer}.
+  */
+public class SparkDataTracer implements DataTracer, Externalizable {
+
+  private final DataTracer dataTracer;
+
+  /**
+   * Constructor. It delegates {@link DataTracer} operations to the current {@link SparkRuntimeContext}.
+   */
+  public SparkDataTracer(DataTracer dataTracer) {
```
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/preview/SparkDataTracer.java]: This is not what I meant. You still need to take `SparkRuntimeContext` through the constructor, otherwise you won't be able to deserialize correctly. You also need to memorize the data tracer name during serialization.

``` java
public class SparkDataTracer implements DataTracer, Externalizable {
  private transient DataTracer dataTracer;

  public SparkDataTracer() {
    // default constructor is always needed for deserialization
  }

  // DefaultSparkExecutionContext call this constructor.
  public SparkDataTracer(SparkRuntimeContext context, String tracerName) {
    this.dataTracer = context.getDataTracer(tracerName);
  }

  @Override
  public void writeExternal(ObjectOutput out) throws IOException {
    out.writeUTF(tracer.getName());
  }

  @Override
  public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
    String tracerName = in.readUTF();
    this.dataTracer = SparkRuntimeContextProvider.get().getDataTracer(tracerName);
  }
}
```

Comment (yaojiefeng) [reply cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/preview/SparkDataTracer.java]: Thanks for the explanation! Did not know the usage of default constructor during deserialization before. So we actually are getting new `DataTracer` instance through SparkRuntimeContext during deserialization. Other examples, like `SparkSecureStore`, is actually getting the `SparkRuntimeContext` in default constructor. So no serialization is actually happening for the delegation instance. Is this correct?

**Comment by chtyim** [comment]
This change LGTM. We need more test cases for the usage from the program context, especially one for Spark to verify that the serialization of `DataTracer` is done correctly.

**Comment by yaojiefeng** [comment]
Cannot rebase due to the first two merge commits, open #7101 to squash and merge the PR
